Milestone emails

aileen opened the pull request:
refs https://www.notion.so/ghost/Marketing-Milestone-email-campaigns-1d2c9dee3cfa4029863edb16092ad5c4

Added milestone email service behind a flag. The service will currently run on boot and is meant to be scheduled soon, which should happen in the next step. For now it's protected behind the alpha flag.

aileen committed:
Fixed JSDoc issues and updated config dep to not require using getters

aileen committed:
Receive default currency from queries

aileen committed:
First pass on hooking up milestone-emails as service behind a flag

allouis commented on ghost/milestone-emails/lib/MilestonesEmailService.js:
```diff
@@ -2,24 +2,19 @@ const Milestone = require('./Milestone');
  */
 
 /**
- * @typedef {object} IMilestoneRepository
```
Same here!

aileen replied:
Not sure I'm following. Are you saying it should be lowercase, or that we're not using the `IMilestoneRepository`, cause I thought we need it for [here](https://github.com/TryGhost/Ghost/pull/16241/files#diff-6b0c2622bbe1fd366bbeb1262f7e2b878284878ad8aea79d15f1b063d24ebe34R27)

allouis replied:
Oh sorry - I meant about the `object` :)

allouis commented on ghost/milestone-emails/lib/MilestonesEmailService.js:
```diff
@@ -115,10 +117,10 @@ module.exports = class MilestonesEmailService {
     /**
      *
-     * @param {Array} goalValues
+     * @param {Array.<number>} goalValues
```
I think `number[]` is preferable here ☺️

aileen replied:
This errors, because I am returning a number. Returning the value of the first index of the filtered Array 😅

allouis reviewed (commented):
Generally we don't want to use the wrapper types like `Object` `Number` `Array` etc...

In the case of `Object` this is actually a lot looser than you think, this would allow a `Number` or an `Array` to be passed, because they are both instances of `Object`.

Instead we wanna use like `object` or `<type>[]` (e.g. `number[]` or `Thing[]`) for a list of types.

I think it's better that we move the labs check into the job that runs, because otherwise switching the labs flag will require a restart of the server for it to take effect

allouis commented on ghost/milestone-emails/lib/MilestonesEmailService.js:
```diff
@@ -2,24 +2,19 @@ const Milestone = require('./Milestone');
 
 /**
  * @template Model
- * @typedef {object} Mention<Model>
```
Actually - what is this type even used for?

aileen replied:
Not used. Removed it. Was a leftover from merging the repository into the API 😅

aileen commented on ghost/milestone-emails/lib/InMemoryMilestoneRepository.js:
```diff
@@ -1,10 +1,9 @@
  */
 
 /**
- * @implements {IMilestoneRepository}
```
It errors for me as the MilestoneAPI doesn't exist anymore and we don't import it as we are in the `IMilestoneRepository`?

aileen committed:
Typedef updates

aileen committed:
More type updates

aileen committed:
Added job files for ARR and members milestones

- The jobs need to be scheduled to run once a week
- Current issue is that I can't access labs to see if the alpha flag is set

aileen commented on ghost/core/core/server/services/milestone-emails/index.js:
```diff
@@ -0,0 +1,71 @@
+module.exports = async (milestoneType) => {
+    // const labs = require('../../../shared/labs');
+
+    // if (labs.isSet('milestoneEmails')) {
```
@allouis I can't seem to access labs. It would always return `false` so not sure how to solve that.

aileen replied:
Note: I also tried requiring it in the job files with the same result

allouis replied:
Do you have `enableDeveloperExperiments` set? That's needed for alpha flags - also can you look at the `labs` setting and see what the value is

aileen replied:
Yup, all set. None of the flags are actually available at this stage it seems 😔

aileen commented on ghost/core/core/server/services/milestone-emails/index.js:
```diff
@@ -0,0 +1,71 @@
+        // TODO: do we need to check if Stripe is live enabled?
+    });
+
+    return await milestonesEmailService.checkMilestones(milestoneType);
```
Not sure if this is the right way to go, because when using it as in-memory storage it gets instantiated every time again, which doesn't make sense I think....

aileen commented on ghost/core/core/server/services/milestone-emails/index.js:
```diff
@@ -0,0 +1,71 @@
+        jobsService.addJob({
+            at: `${s} ${m} ${h} * * ${wd}`, // Every week
+            // at: '55 * * * * *', // every minute for local development
+            job: path.resolve(__dirname, 'jobs/run-arr-milestones.js'),
```
In order to schedule it with a CRON tab I had to remove `offloaded:false` as this seems to be only working for one-off jobs, unless I'm missing something.

The only way then I could make it work was to create job files that immediately invoke, similar to how it's done for `email-analytics` and `update-check` services.

aileen committed:
Removed job scheduling for now

aileen committed:
Change to always use the highest value as default currency

aileen committed:
Fixed logical error for creating milestones when no milestone was hit

aileen committed:
First pass on e2e test for milstone email serice

aileen committed:
Added basic unit tests for Milestone service and queries

aileen committed:
Mock knex for MilestoneQueries unit test

aileen committed:
Removed unneded additional currency check

aileen committed:
Removed unneded model require from unit tests

aileen committed:
Ensure we only fetch ARR milestones when Stripe is configured in live mode

aileen commented on ghost/core/test/e2e-server/services/milestone-emails.test.js:
```diff
@@ -0,0 +1,222 @@
+
+    beforeEach(function () {
+        labsStub = sinon.stub(labs, 'isSet').returns(true);
+        stripeModeStub = sinon.stub(stripeService.api, 'mode').get(() => 'live');
```
I think I messed up the DB tests with this one

aileen committed:
Switched to mockAgent for labs test and rewire to stub StripeService as it was messing with other tests

aileen committed:
More test coverage

aileen committed:
Updated test to catch another case

allouis commented on ghost/core/core/server/services/milestone-emails/index.js:
```diff
@@ -0,0 +1,57 @@
+// Stubbing stripe in test was causing issues. Moved it
```
Can you rename this file to `service.js` and then create an `index.js` with contents of `module.exports = require('./service.js')`

allouis reviewed (commented):
One comment den merge away!

aileen committed:
Wrapped service in index file

aileen committed:
Fixed coverage issues

- Stubbing out the stripe service to mock Stripe live enabled by using `sinon.stub` or `rewire` causes the coverage to decrease massively for no known reason
- I think we should probably create a settingsCache property for this in order to achieve this
